File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException

fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

/**
 *  A function when given a list of strings and a predicate, returns the lowercased
 *  version of the first string satisfying the given predicate. Otherwise, returns null.
 *
 * @param strList A list of strings.
 * @param predicate A predicate where given a string, returns a boolean value.
 * @return First string in strList that satisfies the predicate in lowercase, null otherwise.
 */
fun firstThenLowerCase(strList: List<String>, predicate: (String) -> Boolean): String? {
    return strList.firstOrNull(predicate)?.lowercase()
}

/**
 *  A class that can be optionally initialized with a string that possesses a
 *  read-only property 'phrase' which returns the concatenated string from
 *  current and previous helper method calls.
 *
 * @param sentence An optional string to start the sentence.
 * concatenates the passed argument onto the accumulated sentence.
 * @return The final accumulated sentence, empty string if no arguments
 * were passed.
 */
class say(private val sentence: String? = null) {
    /**
     *  The helper method of say class which concatenates passed string
     *  onto the accumulated sentence.
     *
     *   @param nextWord A string passed into the helper method and() which
     *   @return An instance of the say class with the nextWord added onto
     *   the accumulated sentence. If the sentence is null, initiate the class
     *   with nextWord.
     */
    fun and(nextWord: String): say {
        return say( if(sentence == null) nextWord else "$sentence $nextWord" )
    }

    val phrase: String
        get() = sentence ?: ""
}

/**
 *  A function when given a file name, returns the number of lines that fulfill a certain set of rules.
 *  If the file is not found, throws an IO exception.
 *
 *  Rules:
 *  1) Line is not empty
 *  2) Line is not entirely made of whitespaces
 *  3) The first non-whitespaces character of the line is not '#'
 *
 * @param fileName Name of the file.
 *
 * @return The amount of 'meaningful' lines.
 */
fun meaningfulLineCount(fileName: String): Long {
    var meaningfulCount: Long = 0L
    val fileReader: BufferedReader?
    try {
        fileReader = File(fileName).bufferedReader()
        print(fileReader.readLine())
    } catch (e: IOException) {
        throw IOException("No such file")
    }

    fileReader.useLines {
        lines -> lines.forEach  {
            val trimmedLine: String = it.trim()
            if (trimmedLine.isNotEmpty() && trimmedLine.substring(0, 1) != "#") {
                meaningfulCount++
            }
        }
    }
    return meaningfulCount
}

/**
 *  Represents a quaternion: a four-value representation of three-dimensional rotation given four Double values.
 */
data class Quaternion(
    val a: Double,
    val b: Double,
    val c: Double,
    val d: Double) {
    /**
     *  Static members of the Quaternion dataclass that represent very common quaternions.
     */
    companion object {
            val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
            val I = Quaternion(0.0, 1.0, 0.0, 0.0)
            val J = Quaternion(0.0, 0.0, 1.0, 0.0)
            val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }

    /**
     *  Adds two Quaternions.
     *
     *  @return the resulting Quaternion.
     */
    operator fun plus(other: Quaternion): Quaternion {
            return Quaternion(this.a + other.a, this.b + other.b, this.c + other.c, this.d + other.d)
    }

    /**
     *  Multiplies two quaternions.
     *  Formula courtesy of https://stackoverflow.com/questions/19956555/how-to-multiply-two-quaternions
     *
     *  @return the resulting Quaternion
     */
    operator fun times(other: Quaternion): Quaternion {
        return Quaternion(
            this.a * other.a - this.b * other.b - this.c * other.c - this.d * other.d,  // 1
            this.a * other.b + this.b * other.a + this.c * other.d - this.d * other.c,  // i
            this.a * other.c - this.b * other.d + this.c * other.a + this.d * other.b,  // j
            this.a * other.d + this.b * other.c - this.c * other.b + this.d * other.a  // k
        )
    }

    /**
     *  Returns a list of the Quaternion coefficients in the given order.
     */
    fun coefficients(): List<Double> {
        return listOf(this.a, this.b, this.c, this.d)
    }

    /**
     *  Negates i,j,k values of the Quaternion and returns a new one.
     */
    fun conjugate(): Quaternion {
        return Quaternion(this.a, -this.b, -this.c, -this.d)
    }

    /**
     *  Returns a mathematical representation of the quaternion, ignoring any zero-coefficients.
     */
    override fun toString(): String {
        if (this == ZERO) return "0" // Special case

        val builder = StringBuilder()

        fun symbolFormat(value: Double, symbol: String) {
            when (value) {
                0.0 -> {} // Don't do anything when value is 0
                1.0 -> {if (builder.isNotEmpty()) builder.append("+$symbol") else builder.append(symbol) }
                -1.0 -> builder.append("-$symbol")
                else -> {
                    if (value > 0.0 && builder.isNotEmpty())  builder.append("+")
                    builder.append("$value$symbol")
                }
            }
        }

        if (this.a != 0.0) builder.append(this.a)

        symbolFormat(this.b, "i")
        symbolFormat(this.c, "j")
        symbolFormat(this.d, "k")

        return builder.toString()
    }
}

/**
 *  A sealed interface used to implement empty and node binary search trees nested inside.
 *
 *  @property size returns the size of the current tree.
 *  @property contains traverses through the tree and returns if the lookup value exists.
 *  @property insert adds a node to the right part of the tree.
 *  @property toString represents the binary search tree in string format.
 */
sealed interface BinarySearchTree {
    fun size(): Int
    fun contains(value: String) : Boolean
    fun insert(value: String): BinarySearchTree
    override fun toString(): String

    object Empty : BinarySearchTree {
        override fun size(): Int = 0
        override fun contains(value: String): Boolean = false
        override fun insert(value: String): BinarySearchTree = Node(value, this, this)
        override fun toString(): String = "()" // Test cases expect and empty parentheses for an empty tree
    }

    data class Node(
        private val value: String,
        private val left: BinarySearchTree = Empty,
        private val right: BinarySearchTree = Empty
    ) : BinarySearchTree {
        override fun size(): Int = 1 + left.size() + right.size()

        override fun contains(value: String): Boolean {
            return when {
                // Assuming comparable values
                value == this.value -> true // Found value
                value < this.value -> left.contains(value) // Continue searching to the left
                else -> right.contains(value) // Continue searching to the right
            }
        }

        override fun insert(value: String): BinarySearchTree {
            return when {
                value == this.value -> this // Duplicate entry
                value < this.value -> copy(left = left.insert(value))
                else -> copy(right = right.insert(value))
            }
        }

        override fun toString(): String {
            return "($left$value$right)".replace("()", "")
        }
    }
}

File: haskell/Exercises.hs
module Exercises
    ( change,
      firstThenApply,
      powers,
      meaningfulLineCount,
      Shape(..),
      BST(Empty),
      volume,
      surfaceArea,
      size,
      contains,
      insert,
      inorder
    ) where

import qualified Data.Map as Map
import Data.Text (pack, unpack, replace)
import Data.List(isPrefixOf, find)
import Data.Char(isSpace)

change :: Integer -> Either String (Map.Map Integer Integer)
change amount
    | amount < 0 = Left "amount cannot be negative"
    | otherwise = Right $ changeHelper [25, 10, 5, 1] amount Map.empty
        where
          changeHelper [] remaining counts = counts
          changeHelper (d:ds) remaining counts =
            changeHelper ds newRemaining newCounts
              where
                (count, newRemaining) = remaining `divMod` d
                newCounts = Map.insert d count counts

{- | Returns the result of a given function applied to the first item in the given array that satisfies the given
 predicate.
-}
firstThenApply :: [a] -> (a -> Bool) -> (a -> b) -> Maybe b
firstThenApply xs predicate func = func <$> find predicate xs

-- Generates an infinite sequence of powers of a given base.
powers :: Integral x => x -> [x]
powers base = map (base^) [0..]

{- Returns the number of lines in the given file that are (1) not empty, (2) not all whitespace, and (3) whose first
 character is not '#'.
-}
meaningfulLineCount :: FilePath -> IO Int
meaningfulLineCount filePath = do
    document <- readFile filePath
    let allWhiteSpace = all isSpace
        trimStart = dropWhile isSpace
        isMeaningful line =
            not (allWhiteSpace line) &&
            not ("#" `isPrefixOf` (trimStart line))
    return $ length $ filter isMeaningful $ lines document

-- A type that supports calculations for its volume and surface area.
data Shape
    = Box  Double Double Double
    | Sphere Double
    deriving (Eq, Show)

-- Returns the volume of the given shape.
volume :: Shape -> Double
volume (Box w l d) = w * l * d
volume (Sphere radius) = (4 / 3) * pi * (radius ^ 3)

-- Returns the surface area of the given shape.
surfaceArea :: Shape -> Double
surfaceArea (Box w l d) = 2 * ((w * l) + (w * d) + (l * d))
surfaceArea (Sphere radius) = 4 * pi * (radius ^ 2)

-- A data type that represents an empty or non-empty node in a generic binary search tree.
data BST a = Empty | Node a (BST a) (BST a)
    deriving (Eq)

-- Inserts a new value into a BST.
insert :: Ord a => a -> BST a -> BST a
insert x Empty = Node x Empty Empty
insert x (Node y left right)
    | x < y     = Node y (insert x left) right
    | x > y     = Node y left (insert x right)
    | otherwise = Node y left right

-- Containment check for a BST.
contains :: Ord a => a -> BST a -> Bool
contains _ Empty = False
contains x (Node y left right)
    | x < y     = contains x left
    | x > y     = contains x right
    | otherwise = True

-- Returns the size of a BST.
size :: BST a -> Int
size Empty = 0
size (Node _ left right) = 1 + size left + size right

-- Returns a list of the elements in a BST in order.
inorder :: BST a -> [a]
inorder Empty = []
inorder (Node x left right) = inorder left ++ [x] ++ inorder right

-- Returns a string representation of a BST.
instance (Show a) => Show (BST a) where
    show :: Show a => BST a -> String
    show Empty = "()"
    show (Node x left right) =
        let full = "(" ++ show left ++ show x ++ show right ++ ")" in
        unpack $ replace (pack "()") (pack "") (pack full)